test/K8sServices: Fix externalTrafficPolicy=Local with kube-proxy (on GKE)

gandro opened the pull request:
This commit fixes the test harness for `externalTrafficPolicy=Local` in
the case where we are accessing a node IP without a local backend from a
node with host reachable services enabled.

This is a known incompatibility between our kube-proxy replacement and
upstream kube-proxy. The existing test harness assumed that we only need
to handle this case if we are running without kube-proxy. This
assumption however is wrong, as we are running these tests in hybrid
mode, where we are running with both kube-proxy and Cilium's kube-proxy
replacement.

This has not been hit in our existing test suites up until recently,
because we did not have a test setup with both kube-proxy and the
kube-proxy replacement enabled at the same time. As GKE has been
upgraded to Linux 4.19, it is now using the described hybrid setup which has
caused the tests to break.

The test matrix as of writing now looks as follows:

| Test Suite                 | kube-proxy | kube-proxy replacement  |
| -------------------------- | ---------- | ----------------------- |
| K8s-1.18-kernel-4.9        | Yes        | No  (Kernel 4.9.x)      |
| K8s-1.17-Kernel-4.19       | No         | Yes (Kernel 4.19.57)    |
| K8s-1.12-Kernel-netnext    | No         | Yes (Kernel 5.8.0-rc1+) |
| Cilium-PR-Ginkgo-Tests-k8s | Yes        | No  (Kernel 4.9.x)      |
| Cilium-PR-K8s-GKE          | Yes        | Yes (Kernel 4.19.112+)  |

Fixes: 67f85e3 ("tests: enable additional externalTrafficPolicy=Local tests")

gandro committed:
test/K8sServices: Fix externalTrafficPolicy=Local with kube-proxy

This commit fixes the test harness for `externalTrafficPolicy=Local` in
the case where we are accessing a node IP without a local backend from a
node with host reachable services enabled.

This is a known incompatibility between our kube-proxy replacement and
upstream kube-proxy. The existing test harness assumed that we only need
to handle this case if we are running without kube-proxy. This
assumption however is wrong, as we are running these tests in hybrid
mode, where we are running with both kube-proxy and Cilium's kube-proxy
replacement.

This has not been hit in our existing test suites up until recently,
because we did not have a test setup with both kube-proxy and the
kube-proxy replacement enabled at the same time. As GKE has been
upgraded to Linux 4.19, it is now using the described setup which has
caused the tests to break.

The test matrix as of writing now looks as follows:

| Test Suite                 | kube-proxy | kube-proxy replacement  |
| -------------------------- | ---------- | ----------------------- |
| K8s-1.18-kernel-4.9        | Yes        | No  (Kernel 4.9.x)      |
| K8s-1.17-Kernel-4.19       | No         | Yes (Kernel 4.19.57)    |
| K8s-1.12-Kernel-netnext    | No         | Yes (Kernel 5.8.0-rc1+) |
| Cilium-PR-Ginkgo-Tests-k8s | Yes        | No  (Kernel 4.9.x)      |
| Cilium-PR-K8s-GKE          | Yes        | Yes (Kernel 4.19.112+)  |

Fixes: 67f85e3e06f4 ("tests: enable additional externalTrafficPolicy=Local tests")

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
docs: Update versions used in GKE pipeline

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

brb commented on test/k8sT/Services.go:
```diff
@@ -1026,20 +1029,27 @@ var _ = Describe("K8sServicesTest", func() {
-			// kube-proxy free case since we'll hit the wildcard rule in bpf_sock
+			// In-cluster connectivity from k8s2 to k8s1 IP will still work with
+			// HostRechableServices (regardless of if we are running with or
+			// without kube-proxy) since we'll hit the wildcard rule in bpf_sock
```
Could you extend the comment saying why it won't work on kube-proxy?

gandro replied:
Added the following comment:
```
// This is a known incompatibility with kube-proxy:
// kube-proxy 1.15+ will only load-balance requests from k8s1 to k8s1,
// but not from k8s2 to k8s1. In the k8s2 to k8s1 case, kube-proxy
// would send traffic to k8s1, where it would be subsequently
// dropped, because k8s1 has no service backend.
// However, if HostReachableServices is enabled, then Cilium does
// the service translation already on the client node, bypassing
// kube-proxy completely.
```

brb reviewed (approved):
Thanks!

nebril committed:
ci: Delete cilium-node-init ds before cilium install

`--validate` helm flag introduced in 82cc7c3d07 caused ci to fail in gke
where we enable node init daemonset. It needs to be cleared before
cilium installation.

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
